Rust bundle patch introduction

abaez opened the pull request:
First patch starting from 'initial commit' 1d8df75 to 754a618 in
[howl-rust].

These changesets introduce the [Rust] bundle into howl with full
lexical definition. The changesets continue feature request #224 and
also allow the progress of the bundle, with easier integration into
howl, from the [howl-rust] source.

[howl-rust]: https://bitbucket.org/a_baez/howl-rust
[Rust]: https://www.rust-lang.org/en-US/

kirbyfan64 commented on bundles/rust/LICENSE:
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2016 Alejandro Baez https://twitter.com/a_baez
```
FWIW, normally there's "just" the [copyright stanza](https://github.com/howl-editor/howl/blob/master/bundles/pascal/init.moon#L1-L2) at the top of each file ([which I usually forget](https://github.com/howl-editor/howl/pull/181/commits/5c331574657a683ac28198b18a4ff3cc5a940376)).

kirbyfan64 commented on bundles/rust/init.moon:
```diff
@@ -0,0 +1,25 @@
+
+mode_reg =
+  name: 'rust'
+  aliases: 'rust'
```
Maybe `rs` could be an alias?

kirbyfan64 commented on bundles/rust/rust_mode.moon:
```diff
@@ -0,0 +1,29 @@
+
+import style from howl.ui
+
+style.define 'extension', "special"
```
Is there really a need for this? I think these might be global definitions, which could interfere with other modes.

kirbyfan64 commented on bundles/rust/rust_lexer.moon:
```diff
@@ -0,0 +1,93 @@
+  operator = c 'operator', S'+-/*%<>!=`^~@&|?#~:;,.()[]{}'
+
+  -- Attributes.
+  attribute = c 'attribute', (P'#![' + P'#[') * scan_until(eol + P']')
```
This confuses me a little. For starters, it won't highlight the closing brace (I think using `scan_to` instead of `scan_until` will fix that). But...why is the `eol +` there? The `eol` will always match, so lpeg will never even bother to try the `P']'`. My guess is that you were trying to highlight it correctly, even if the attribute was still being typed; in that case, I think `P']' + eol` would work better.

kirbyfan64 commented on bundles/rust/rust_mode.moon:
```diff
@@ -0,0 +1,29 @@
+    '"': '"'
+  }
+
+  default_config:
```
I don't think this is needed, since that's pretty much the standard, default config.

kirbyfan64 commented:
Nice! As an aside, maybe this could highlight stuff like `struct Xyz` as a `type_def`?

See [here](https://github.com/howl-editor/howl/blob/master/bundles/python/python_lexer.moon#L55) for an example.

its-suun commented:
I haven't tested your bundle out but from source I can say that some things get wrong(or not) highlighted:
- `_123` gets highlighted (numbers don't start with _)
- `123i32` or `123.42f64` doesn't get highlighted
- `0b_101` gets highlighted (doesn't start with _ after `0b`)
- `r#"abc"#` doesn't get fully highlighted (only from # till end), same with chars
- `br##" abc"##` doesn't get fully highlighted ([reference](https://doc.rust-lang.org/reference.html#raw-string-literals))

These are really minor issues since they are rarely in rust code. Anyways thanks for working on the mode!

abaez commented:
Hey @kirbyfan64, 

Thank you so much for doing a quick review! Now to answer your questions in the order received.

> https://github.com/howl-editor/howl/pull/257#discussion-diff-75767500

This is due to my old ldoc/lua habit. Most of the time, I just have a generator for each time I make a new file on my personal projects. No issue here in replacing the headers. :+1: 

> https://github.com/howl-editor/howl/pull/257#discussion-diff-75769160

Didn't even notice I had a double reference to **rust**. Fixed!

> https://github.com/howl-editor/howl/pull/257#discussion-diff-75769651

Yes, it really is necessary. Style definition aliasing is number two( one being PEG definitions) in being able to manipulate the text editor's lexer to work appropriately for a language. If the style definition are hardcoded only to standard `ansi_c`, it makes the lexer (at least in my opinion) harder to appropriate on how you want a style defined. 

With that said, the [specsheet written for howl][1] states `howl.ui.style.define` as mutable. Also, if I'm not mistaken, I doubt the `style.define` would apply to other modes considering **Moonscript**'s local scope by default state and the appropriate mode call for the `rust_mode`. If this is _not_ the case, then I know what my first RFC for Howl will be. 

> https://github.com/howl-editor/howl/pull/257#discussion-diff-75770295

You're totally right. I don't know what I was thinking here. Fixed!

> https://github.com/howl-editor/howl/pull/257#discussion-diff-75770874

The `auto_pair` is almost the same, but not quite. Rust has a tendency to use *single quotes* for user defined scoped memory allocation's reference; [lifetimes][2]. That means any time a `'` is used for a [lifetime][2], an auto pair would appear for it, instead of just a singular `'`. With making sure that the `auto_pair` does not use single quotes for auto pairs, this annoyance is avoided.

> https://github.com/howl-editor/howl/pull/257#issuecomment-241571044

I have to do more research, but I think I'll probably make something of the sort with `impl`, `enum`, and `struct` like you stated. 

---

Thanks again for all the details. Greatly appreciate the help in getting this little bundle's initial patch on Howl!

[1]: http://howl.io/doc/spec/ui/style_spec.html
[2]: https://doc.rust-lang.org/book/lifetimes.html

kirbyfan64 commented:
> Yes, it really is necessary. Style definition aliasing is number two( one being PEG definitions) in being able to manipulate the text editor's lexer to work appropriately for a language.

Huh? Can't you just use the style definitions directly? e.g.:

```
attribute = c 'preproc', ...
```

Since the variable's named `attribute`, it's already self-evident that this is highlighting attributes; if anything the explicitness helps readability. When I first tried this mode out and noticed the highlighting issue with attributes, I did a ctrl+f for `preproc` and was puzzled when I didn't find it.

> If the style definition are hardcoded only to standard `ansi_c`

...sorry, but I have no clue what you're referring to. What do you mean by `ansi_c`?

Also, a quick look at the source shows that `style.define` really is global.

> The `auto_pair` is almost the same...

I was referring to the `default_config` part....

abaez commented:
One note. Please don't accept the patch until it's rebased. I do want to clean this mess later on. For now, I'm simply adding changes as requested to keep the referenced comment till then.

As for @its-suun: 

* I'll work on this. 
* This I have no idea how to fix. On Textadept, the lpeg_lexer (which is the same piece from textadept used in howl), would read both the type and number lexical definition. Sounds like a bug for me to fix here. :+1:  
* numbers I didn't translate exactly. That's my mistake. Will work on this.
* same as numbers.
* same as numbers.

abaez commented:
Hey @kirbyfan64, 

> Huh? Can't you just use the style definitions directly? 

That was the whole point I was trying to avoid, when I stated to _not_ define by `ansi_c`. By `ansi_c`, I was referring to [ANSI C of textadept][1] where `lpeg_lexer` originates from. If it has to be hard coded, then it's no big deal. Just a slight disappointment of the `lpeg_lexer` used here. :(

> I was referring to the default_config part....

Ha! Sorry about that. I thought you were referencing `auto_pair`. With that said, then yes, I still do think it (`default_config`) is needed. Simply from personal anecdotal view, I use only 2 spaces as default with **79 characters** for the limit. I don't want to remember each and every time I open up a rust file to change my defaults. Already have enough to deal with than having to remember the editor's "default" preferences for all lexers.  

[1]: http://foicica.com/hg/scintillua/file/1c8cf7032633/lexers/ansi_c.lua

kirbyfan64 commented:
> I don't want to remember each and every time I open up a rust file to change my defaults. Already have enough to deal with than having to remember the editor's "default" preferences for all lexers.

http://howl.io/doc/manual/configuration.html#setting-variables-upon-startup

So you could just do:

```moonscript
howl.mode.configure 'c', {
  indent: 4
  -- ...
}
```

kirbyfan64 commented:
and put that in your `init.moon`.

abaez commented:
From @kirbyfan64: 
> So you could just do: ...

Yes that  can work, but don't think user definition applies here. The default configuration is meant for the specific's bundle formatting requirements. In Rust's case, the default or at least recommended format is [4 spaces with 99 characters limit][1]. While I completely understand in leaving out preferences, as how I represented this issue prior, the style guideline is standard formatting on Rust. I can't seem to reason how it would be better to remove the guideline, when vim, emac, textadept, sublime text, and etc all use the same setting by default.

[1]: https://github.com/rust-lang/rust/blob/master/src/doc/style/style/whitespace.md

nilnor commented:
Re styles: A bundle can define it's own styles, in order to facilitate specific styling later in user or theme files. In those cases `define_default` is a better option as it does not override any existing definition, should one exist. When defining specific styles I think it's better to name space them , in this case `extension` is rather generic. The CSS bundle is one example where custom styles are used via aliases: https://github.com/howl-editor/howl/blob/master/bundles/css/css_lexer.moon.

nilnor commented:
Re the edge column it makes sense to configure that appropriately for Rust. The issue right might be that there's only one variable for configuring it, which means setting a specific column value also means that it should be shown. So a user who does not want to see the edge line will now do so, unless they're manually overriding the setting. Perhaps we should introduce a new option, `show_edge_line`, for controlling whether the edge line is shown and have `edge_column` only be the column. That seems like it work out better, what do you think?

abaez commented:
Hey @nilnor, 

> Re styles: A bundle can define it's own styles, in order to facilitate specific styling later in user or theme files. ...

Thanks for the input on `define_default`, I was mixed up with which was global and local to the bundle, for defined custom styles. A small question on this but why is `define` the global state and not `define_default` method? From variable definition perspective, I would assume `define_default` would be the global scoped method instead of `define`. Anyway, I'll get on in changing to a namespace structure for the `.define_default` of Rust. :)

>  Perhaps we should introduce a new option, `show_edge_line` ...

Definitely agree. I think we can do one step better, by borrowing a little and doing [`edge_mode` with the `edge_column`][1]. Let me know what you think, don't want to get too offtopic, so I'll make an RFC momentarily for it. ;)

---

Thanks again to @nilnor and @kirbyfan64. I really do appreciate the help here.

[1]: http://foicica.com/textadept/api.html#buffer.edge_column

nilnor commented:
@abaez both `define` and `define_default` are global, the difference is that `define_default` only defines the style using the specified definition unless it's already defined. So there are no "local" styles. Using aliases it's possible to define e.g. `rust_extension` with a reasonable alias that makes it blend in well with all kind of themes. Should themes or a user later want to customize this further they can then specifically customize the `rust_extension` style. Styles need to work with all kind of themes and preferences, so having anything more "local" than that does not really make sense as I see it, i.e. a bundle can not possibly provide a style definition that will work well with all imaginary themes.

Re the edge mode: yes, having the ability to draw it in two ways could be done (saw the PR).

abaez commented:
I've gone ahead and taken everything discussed here. Applied all the changes which should make the patch acceptable for merger. I've also rebased the changesets to one single commit for ease in integration. 

Please let me know if there's an issue here, else I think the patch is ready for introduction into howl.

nilnor commented on bundles/rust/README.md:
```diff
@@ -0,0 +1,33 @@
+# howl-rust
```
This README has not been updated to account for inclusion in core - the background history (while interesting), is not relevant for a core bundle IMO, and the `usage` section is obviously not applicable once it's merged.

nilnor commented on bundles/rust/init.moon:
```diff
@@ -0,0 +1,21 @@
+-- Copyright 2012-2016 The Howl Developers
```
Nitpicking, but as this bundle was created this year the copyright should say "Copyright 2016".

abaez replied:
Sorry about that, I was only going with that @kirbyfan64 spoke upon in setting for the Copyright claim: https://github.com/howl-editor/howl/pull/257#discussion_r75767500

I'll change this!

nilnor commented:
@abaez Some minor format nitpicking from me :) I'll leave Rust specific things for others to comment on, but barring further objections I'd say we can merge after the formatting fixes.

abaez commented:
Hey @nilnor, 

I went ahead and removed the `{LICENSE,README}.md` of this version. I've also changed the copyright to the correct start year of the bundle, as request. :+1:  Let me know anything else you may want to get this little bundle upstream.

nilnor commented:
Sorry for the delay, I was away for a week. In it goes!
